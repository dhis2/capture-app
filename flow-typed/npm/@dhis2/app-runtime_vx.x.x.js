// flow-typed signature: 16b8349e118448457b6534f0314306fa
// flow-typed version: <<STUB>>/@dhis2/app-runtime_v^2.6.1/flow_v0.132.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@dhis2/app-runtime'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@dhis2/app-runtime' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@dhis2/app-runtime/build/cjs/lib' {
  declare module.exports: any;
}

declare module '@dhis2/app-runtime/build/es/lib' {
  declare module.exports: any;
}

// Filename aliases
declare module '@dhis2/app-runtime/build/cjs/lib.js' {
  declare module.exports: $Exports<'@dhis2/app-runtime/build/cjs/lib'>;
}
declare module '@dhis2/app-runtime/build/es/lib.js' {
  declare module.exports: $Exports<'@dhis2/app-runtime/build/es/lib'>;
}
