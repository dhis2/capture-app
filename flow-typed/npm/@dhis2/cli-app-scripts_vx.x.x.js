// flow-typed signature: 522c3cdf2366afe55a40a93d66b99a9e
// flow-typed version: <<STUB>>/@dhis2/cli-app-scripts_v^4.0.9/flow_v0.132.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@dhis2/cli-app-scripts'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@dhis2/cli-app-scripts' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@dhis2/cli-app-scripts/config/app.babel.config' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/config/babel.config' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/config/d2.config.app' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/config/d2.config.lib' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/config/init.App.test' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/config/init.entrypoint' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/config/jest.config' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/config/jest.file.mock' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/config/jest.identity.mock' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/config/jest.styled-jsx-css.mock' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/config/jest.transform' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/config/rollup.config' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/commands/build' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/commands/clean' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/commands/i18n' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/commands/init' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/commands/start' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/commands/test' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/bundleApp' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/compiler/compile' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/compiler/compileApp' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/compiler/compileLibrary' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/compiler' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/exitOnCatch' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/generateManifest' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/i18n/extract' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/i18n/generate' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/i18n/helpers' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/i18n' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/i18n/locales' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/i18n/validate' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/loadEnvFiles' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/parseConfig' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/paths' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/shell/bootstrap' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/shell/env' {
  declare module.exports: any;
}

declare module '@dhis2/cli-app-scripts/src/lib/shell' {
  declare module.exports: any;
}

// Filename aliases
declare module '@dhis2/cli-app-scripts/config/app.babel.config.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/config/app.babel.config'>;
}
declare module '@dhis2/cli-app-scripts/config/babel.config.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/config/babel.config'>;
}
declare module '@dhis2/cli-app-scripts/config/d2.config.app.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/config/d2.config.app'>;
}
declare module '@dhis2/cli-app-scripts/config/d2.config.lib.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/config/d2.config.lib'>;
}
declare module '@dhis2/cli-app-scripts/config/init.App.test.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/config/init.App.test'>;
}
declare module '@dhis2/cli-app-scripts/config/init.entrypoint.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/config/init.entrypoint'>;
}
declare module '@dhis2/cli-app-scripts/config/jest.config.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/config/jest.config'>;
}
declare module '@dhis2/cli-app-scripts/config/jest.file.mock.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/config/jest.file.mock'>;
}
declare module '@dhis2/cli-app-scripts/config/jest.identity.mock.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/config/jest.identity.mock'>;
}
declare module '@dhis2/cli-app-scripts/config/jest.styled-jsx-css.mock.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/config/jest.styled-jsx-css.mock'>;
}
declare module '@dhis2/cli-app-scripts/config/jest.transform.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/config/jest.transform'>;
}
declare module '@dhis2/cli-app-scripts/config/rollup.config.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/config/rollup.config'>;
}
declare module '@dhis2/cli-app-scripts/src/commands/build.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/commands/build'>;
}
declare module '@dhis2/cli-app-scripts/src/commands/clean.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/commands/clean'>;
}
declare module '@dhis2/cli-app-scripts/src/commands/i18n.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/commands/i18n'>;
}
declare module '@dhis2/cli-app-scripts/src/commands/init.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/commands/init'>;
}
declare module '@dhis2/cli-app-scripts/src/commands/start.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/commands/start'>;
}
declare module '@dhis2/cli-app-scripts/src/commands/test.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/commands/test'>;
}
declare module '@dhis2/cli-app-scripts/src/index' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src'>;
}
declare module '@dhis2/cli-app-scripts/src/index.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/bundleApp.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/bundleApp'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/compiler/compile.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/compiler/compile'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/compiler/compileApp.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/compiler/compileApp'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/compiler/compileLibrary.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/compiler/compileLibrary'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/compiler/index' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/compiler'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/compiler/index.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/compiler'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/exitOnCatch.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/exitOnCatch'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/generateManifest.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/generateManifest'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/i18n/extract.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/i18n/extract'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/i18n/generate.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/i18n/generate'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/i18n/helpers.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/i18n/helpers'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/i18n/index' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/i18n'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/i18n/index.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/i18n'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/i18n/locales.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/i18n/locales'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/i18n/validate.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/i18n/validate'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/loadEnvFiles.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/loadEnvFiles'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/parseConfig.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/parseConfig'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/paths.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/paths'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/shell/bootstrap.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/shell/bootstrap'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/shell/env.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/shell/env'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/shell/index' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/shell'>;
}
declare module '@dhis2/cli-app-scripts/src/lib/shell/index.js' {
  declare module.exports: $Exports<'@dhis2/cli-app-scripts/src/lib/shell'>;
}
