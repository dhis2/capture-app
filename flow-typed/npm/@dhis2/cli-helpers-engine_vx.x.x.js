// flow-typed signature: 6b7a44e4abb9853a2939e68e6e1e2d32
// flow-typed version: <<STUB>>/@dhis2/cli-helpers-engine_v^2.1.1/flow_v0.132.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@dhis2/cli-helpers-engine'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@dhis2/cli-helpers-engine' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@dhis2/cli-helpers-engine/lib/cache/Cache' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/cache/fetchAndExtract' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/cache' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/cache/middleware' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/configDefaults' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/exec' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/groupGlobalOptions' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/makeEntryPoint' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/namespace' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/notifyOfUpdates' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/prompt' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/reporter' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/tryCatchAsync' {
  declare module.exports: any;
}

declare module '@dhis2/cli-helpers-engine/lib/utils/strings' {
  declare module.exports: any;
}

// Filename aliases
declare module '@dhis2/cli-helpers-engine/index' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine'>;
}
declare module '@dhis2/cli-helpers-engine/index.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine'>;
}
declare module '@dhis2/cli-helpers-engine/lib/cache/Cache.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/cache/Cache'>;
}
declare module '@dhis2/cli-helpers-engine/lib/cache/fetchAndExtract.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/cache/fetchAndExtract'>;
}
declare module '@dhis2/cli-helpers-engine/lib/cache/index' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/cache'>;
}
declare module '@dhis2/cli-helpers-engine/lib/cache/index.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/cache'>;
}
declare module '@dhis2/cli-helpers-engine/lib/cache/middleware.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/cache/middleware'>;
}
declare module '@dhis2/cli-helpers-engine/lib/configDefaults.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/configDefaults'>;
}
declare module '@dhis2/cli-helpers-engine/lib/exec.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/exec'>;
}
declare module '@dhis2/cli-helpers-engine/lib/groupGlobalOptions.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/groupGlobalOptions'>;
}
declare module '@dhis2/cli-helpers-engine/lib/makeEntryPoint.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/makeEntryPoint'>;
}
declare module '@dhis2/cli-helpers-engine/lib/namespace.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/namespace'>;
}
declare module '@dhis2/cli-helpers-engine/lib/notifyOfUpdates.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/notifyOfUpdates'>;
}
declare module '@dhis2/cli-helpers-engine/lib/prompt.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/prompt'>;
}
declare module '@dhis2/cli-helpers-engine/lib/reporter.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/reporter'>;
}
declare module '@dhis2/cli-helpers-engine/lib/tryCatchAsync.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/tryCatchAsync'>;
}
declare module '@dhis2/cli-helpers-engine/lib/utils/strings.js' {
  declare module.exports: $Exports<'@dhis2/cli-helpers-engine/lib/utils/strings'>;
}
