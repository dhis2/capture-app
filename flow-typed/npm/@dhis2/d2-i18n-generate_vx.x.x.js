// flow-typed signature: ea7684ed063ab008f97fe36fbf50e745
// flow-typed version: <<STUB>>/@dhis2/d2-i18n-generate_v^1.1.1/flow_v0.127.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@dhis2/d2-i18n-generate'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@dhis2/d2-i18n-generate' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */


// Filename aliases
declare module '@dhis2/d2-i18n-generate/index' {
  declare module.exports: $Exports<'@dhis2/d2-i18n-generate'>;
}
declare module '@dhis2/d2-i18n-generate/index.js' {
  declare module.exports: $Exports<'@dhis2/d2-i18n-generate'>;
}
