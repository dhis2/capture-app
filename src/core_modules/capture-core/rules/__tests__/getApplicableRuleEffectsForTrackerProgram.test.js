import { variableSourceTypes } from '@dhis2/rules-engine-javascript';
import {
    TrackerProgram,
    ProgramStage,
    RenderFoundation,
    Section,
    DataElement,
    dataElementTypes,
    OptionSet,
    Option,
} from '../../metaData';
import { getApplicableRuleEffectsForTrackerProgram } from '..';

const mockGetProgramRuleEffects = jest.fn().mockImplementation(() => [
    {
        id: 'effectId',
        type: 'DISPLAYTEXT',
        message: 'display effect',
    },
]);

const mockOptionSet = new OptionSet('optionSet1', [new Option('option1', 'opt1')]);
jest.mock('@dhis2/rules-engine-javascript/build/cjs/RulesEngine', () => ({
    RulesEngine: jest
        .fn()
        .mockImplementation(() => ({ getProgramRuleEffects: (...args) => mockGetProgramRuleEffects(...args) })),
}));

jest.mock('../../metaDataMemoryStores/constants/constants.store', () => ({
    constantsStore: { get: () => [{ id: 'constantId1', value: '1' }] },
}));

jest.mock('../../metaDataMemoryStores/optionSets/optionSets.store', () => ({
    optionSetStore: { get: () => [mockOptionSet] },
}));

describe('getApplicableRuleEffectsForTrackerProgram', () => {
    const currentEvent = {
        da1Id: 'currentEventText',
        dueDate: '2021-05-31T09:51:38.134',
        enrollmentId: 'vVtmDlsu3me',
        enrollmentStatus: 'ACTIVE',
        eventDate: '2021-05-31T00:00:00.000',
        eventId: 'BxGzDJK3JqN',
        orgUnitId: 'DiszpKrYNg8',
        orgUnitName: 'Ngelehun CHC',
        programId: 'IpHINAT79UW',
        programStageId: 'A03MvHHogjR',
        status: 'ACTIVE',
        trackedEntityInstanceId: 'vCGpQAWG17I',
    };

    const otherEvents = [
        {
            da1Id: 'otherEventText',
            dueDate: '2021-05-31T09:51:38.134',
            enrollmentId: 'vVtmDlsu3me',
            enrollmentStatus: 'ACTIVE',
            eventDate: '2021-05-31T00:00:00.000',
            eventId: 'BxGzDJK3JqN',
            orgUnitId: 'DiszpKrYNg8',
            orgUnitName: 'Ngelehun CHC',
            programId: 'IpHINAT79UW',
            programStageId: 'A03MvHHogjR',
            status: 'ACTIVE',
            trackedEntityInstanceId: 'vCGpQAWG17I',
        },
    ];

    const orgUnit = { id: 'DiszpKrYNg8', code: 'Ngelehun CHC' };

    const enrollmentData = {
        enrolledAt: '2021-05-31T00:00:00.000',
        enrollmentId: 'vVtmDlsu3me',
        occurredAt: '2021-05-31T00:00:00.000',
    };

    const programStage = new ProgramStage((stage) => {
        stage.id = 'st1Id';
        stage.name = 'stage1';
        stage.programRules = [
            {
                id: 'rule1Id',
                name: 'rule1',
                displayName: 'rule1',
                priority: 1,
                condition: 'true',
                programId: 'IpHINAT79UW',
                programStageId: 'st1Id',
                programRuleActions: [],
            },
        ];

        stage.stageForm = new RenderFoundation((foundation) => {
            const section = new Section((initSection) => {
                initSection.id = 's1Id';
                initSection.name = 'section1';
                const dataElement = new DataElement((initDataElement) => {
                    initDataElement.id = 'da1Id';
                    initDataElement.name = 'dataElement1';
                    initDataElement.type = dataElementTypes.TEXT;
                });
                initSection.addElement(dataElement);
            });
            foundation.addSection(section);
        });
    });

    const program = new TrackerProgram((initProgram) => {
        initProgram.id = 'IpHINAT79UW';
        initProgram.attributes = [
            new DataElement((element1) => {
                element1.id = 'zDhUuAYrxNC';
                element1.name = 'SomeText';
                element1.type = dataElementTypes.TEXT;
            }),
            new DataElement((element2) => {
                element2.id = 'lZGmxYbs97q';
                element2.name = 'someNumber';
                element2.type = dataElementTypes.NUMBER;
            }),
            new DataElement((element3) => {
                element3.id = 'lZGmxYbs96q';
                element3.name = 'SomeDate';
                element3.type = dataElementTypes.DATE;
                element3.optionSet = { id: 'optionSet', name: 'optionSet', code: 'optionSet' };
            }),
            new DataElement((element4) => {
                element4.id = 'w75KJ2mc4zz';
                element4.name = 'SomeOrganinsationUnit';
                element4.type = dataElementTypes.ORGANISATION_UNIT;
            }),
        ];
        initProgram.addStage(programStage);
        initProgram.programRuleVariables = [
            {
                dataElementId: 'UXz7xuGCEhU',
                displayName: 'Test',
                id: 'PUQZWgmQ0jx',
                programId: 'IpHINAT79UW',
                programRuleVariableSourceType: variableSourceTypes.DATAELEMENT_NEWEST_EVENT_PROGRAM,
                useNameForOptionSet: true,
            },
            {
                dataElementId: 'H6uSAMO5WLD',
                displayName: 'apgarcomment',
                id: 'aKpfPKSRQnv',
                programId: 'IpHINAT79UW',
                programRuleVariableSourceType: variableSourceTypes.DATAELEMENT_NEWEST_EVENT_PROGRAM,
                useNameForOptionSet: true,
            },
            {
                dataElementId: 'a3kGcGDCuk6',
                displayName: 'apgarscore',
                id: 'g2GooOydipB',
                programId: 'IpHINAT79UW',
                programRuleVariableSourceType: variableSourceTypes.DATAELEMENT_NEWEST_EVENT_PROGRAM,
                useNameForOptionSet: true,
            },
        ];

        initProgram.programRules = [
            {
                condition: 'true',
                displayName: 'TestRule',
                id: 'JJDQxgHuuL2',
                programId: 'IpHINAT79UW',
                programRuleActions: [
                    {
                        data: '#{Test}',
                        id: 'CQaifjkoFEU',
                        location: 'feedback',
                        programRuleActionType: 'DISPLAYTEXT',
                    },
                ],
            },
        ];
    });

    const attributeValues = {
        zDhUuAYrxNC: 'someTextValue',
        lZGmxYbs97q: 5,
        lZGmxYbs96q: '1995-12-17T03:24:00',
    };

    test('RulesEngine called with computed arguments from getApplicableRuleEffectsForTrackerProgram', () => {
        getApplicableRuleEffectsForTrackerProgram({
            program,
            stage: programStage,
            orgUnit,
            currentEvent,
            otherEvents,
            attributeValues,
            enrollmentData,
        });

        const rulesEngineInput = mockGetProgramRuleEffects.mock.calls[0][0];

        const { optionSets, dataElements, trackedEntityAttributes } = rulesEngineInput;
        expect(Object.keys(optionSets).length).toBe(1);
        expect(Object.keys(dataElements).length).toBe(1);
        expect(Object.keys(trackedEntityAttributes).length).toBe(4);

        const { programRules, constants } = rulesEngineInput.programRulesContainer;
        expect(programRules.length).toBe(2);
        expect(constants.length).toBe(1);

        expect(rulesEngineInput).toMatchSnapshot();
    });

    test('Hierarchical result', () => {
        const effects = getApplicableRuleEffectsForTrackerProgram({
            program,
            stage: programStage,
            orgUnit,
            currentEvent,
            otherEvents,
            attributeValues,
            enrollmentData,
        });

        expect(effects.DISPLAYTEXT).toBeDefined();
    });

    test('Flat result', () => {
        const effects = getApplicableRuleEffectsForTrackerProgram(
            {
                program,
                stage: programStage,
                orgUnit,
                currentEvent,
                otherEvents,
                attributeValues,
                enrollmentData,
            },
            true,
        );

        expect(Array.isArray(effects)).toBe(true);
    });

    test('RulesEngine called without programRules', () => {
        const effects = getApplicableRuleEffectsForTrackerProgram({
            program: new TrackerProgram((initProgram) => {
                initProgram.programRules = [];
            }),
            stage: new ProgramStage((stage) => {
                stage.programRules = [];
            }),
            orgUnit,
            currentEvent,
            otherEvents,
            attributeValues,
            enrollmentData,
        });

        expect(effects).toStrictEqual([]);
    });

    test('currentEvent without a programStageId', () => {
        const effects = getApplicableRuleEffectsForTrackerProgram({
            program,
            stage: programStage,
            orgUnit,
            currentEvent: {},
            otherEvents,
            attributeValues,
            enrollmentData,
        });

        expect(effects.DISPLAYTEXT).toBeDefined();
    });
});
