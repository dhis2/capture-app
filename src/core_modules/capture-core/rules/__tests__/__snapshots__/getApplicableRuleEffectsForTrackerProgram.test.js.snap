// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getApplicableRuleEffectsForTrackerProgram RulesEngine called with computed arguments from getApplicableRuleEffectsForTrackerProgram 1`] = `
Object {
  "currentEvent": Object {
    "da1Id": "currentEventText",
    "dueDate": "2021-05-31T09:51:38.134",
    "enrollmentId": "vVtmDlsu3me",
    "enrollmentStatus": "ACTIVE",
    "eventDate": "2021-05-31T00:00:00.000",
    "eventId": "BxGzDJK3JqN",
    "orgUnitId": "DiszpKrYNg8",
    "orgUnitName": "Ngelehun CHC",
    "programId": "IpHINAT79UW",
    "programStageId": "A03MvHHogjR",
    "programStageName": undefined,
    "status": "ACTIVE",
    "trackedEntityInstanceId": "vCGpQAWG17I",
  },
  "dataElements": Object {
    "da1Id": Object {
      "id": "da1Id",
      "optionSetId": undefined,
      "valueType": "TEXT",
    },
  },
  "optionSets": Object {
    "0": Object {
      "displayName": undefined,
      "id": "optionSet1",
      "options": Array [],
    },
  },
  "otherEvents": Array [
    Object {
      "da1Id": "otherEventText",
      "dueDate": "2021-05-31T09:51:38.134",
      "enrollmentId": "vVtmDlsu3me",
      "enrollmentStatus": "ACTIVE",
      "eventDate": "2021-05-31T00:00:00.000",
      "eventId": "BxGzDJK3JqN",
      "orgUnitId": "DiszpKrYNg8",
      "orgUnitName": "Ngelehun CHC",
      "programId": "IpHINAT79UW",
      "programStageId": "A03MvHHogjR",
      "status": "ACTIVE",
      "trackedEntityInstanceId": "vCGpQAWG17I",
    },
  ],
  "programRulesContainer": Object {
    "constants": Array [
      Object {
        "id": "constantId1",
        "value": "1",
      },
    ],
    "programRuleVariables": Array [
      Object {
        "dataElementId": "UXz7xuGCEhU",
        "displayName": "Test",
        "id": "PUQZWgmQ0jx",
        "programId": "IpHINAT79UW",
        "programRuleVariableSourceType": "DATAELEMENT_NEWEST_EVENT_PROGRAM",
        "useNameForOptionSet": true,
      },
      Object {
        "dataElementId": "H6uSAMO5WLD",
        "displayName": "apgarcomment",
        "id": "aKpfPKSRQnv",
        "programId": "IpHINAT79UW",
        "programRuleVariableSourceType": "DATAELEMENT_NEWEST_EVENT_PROGRAM",
        "useNameForOptionSet": true,
      },
      Object {
        "dataElementId": "a3kGcGDCuk6",
        "displayName": "apgarscore",
        "id": "g2GooOydipB",
        "programId": "IpHINAT79UW",
        "programRuleVariableSourceType": "DATAELEMENT_NEWEST_EVENT_PROGRAM",
        "useNameForOptionSet": true,
      },
    ],
    "programRules": Array [
      Object {
        "condition": "true",
        "displayName": "TestRule",
        "id": "JJDQxgHuuL2",
        "programId": "IpHINAT79UW",
        "programRuleActions": Array [
          Object {
            "data": "#{Test}",
            "id": "CQaifjkoFEU",
            "location": "feedback",
            "programRuleActionType": "DISPLAYTEXT",
          },
        ],
      },
      Object {
        "condition": "true",
        "displayName": "rule1",
        "id": "rule1Id",
        "name": "rule1",
        "priority": 1,
        "programId": "IpHINAT79UW",
        "programRuleActions": Array [],
        "programStageId": "st1Id",
      },
    ],
  },
  "selectedEnrollment": Object {
    "enrolledAt": "2021-05-31T00:00:00.000",
    "enrollmentId": "vVtmDlsu3me",
    "occurredAt": "2021-05-31T00:00:00.000",
  },
  "selectedEntity": Object {
    "lZGmxYbs96q": "1995-12-17T03:24:00",
    "lZGmxYbs97q": 5,
    "zDhUuAYrxNC": "someTextValue",
  },
  "selectedOrgUnit": Object {
    "code": "Ngelehun CHC",
    "id": "DiszpKrYNg8",
  },
  "selectedUserRoles": Array [
    "UYXOT4A7JMI",
    "Ufph3mGRmMo",
    "Euq3XfEIEbx",
    "cUlTcejWree",
    "aNk5AyC7ydy",
    "TMK9CMZ2V98",
    "Ql6Gew7eaX6",
    "Pqoy4DLOdMK",
    "DRdaVRtwmG5",
    "jRWSNIHdKww",
    "txB7vu1w2Pr",
    "xJZBzAHI88H",
    "XS0dNzuZmfH",
  ],
  "trackedEntityAttributes": Object {
    "lZGmxYbs96q": Object {
      "id": "lZGmxYbs96q",
      "optionSetId": undefined,
      "valueType": "DATE",
    },
    "lZGmxYbs97q": Object {
      "id": "lZGmxYbs97q",
      "optionSetId": undefined,
      "valueType": "NUMBER",
    },
    "w75KJ2mc4zz": Object {
      "id": "w75KJ2mc4zz",
      "optionSetId": undefined,
      "valueType": "ORGANISATION_UNIT",
    },
    "zDhUuAYrxNC": Object {
      "id": "zDhUuAYrxNC",
      "optionSetId": undefined,
      "valueType": "TEXT",
    },
  },
}
`;
